docs/plugins/virtual_environment_docs/5-survey.md

```diff
@@ -0,0 +1,115 @@
+    - [Testing widgets](#testing-widgets)  
+    - [Better signal or slot documentation or functionality](#better-signal-or-slot-documentation-or-functionality)  
+    - [Creating or deleting a point or shape layer](#creating-or-deleting-a-point-or-shape-layer)  
+    - [Github tests failing](#github-tests-failing)  
+    - [Cancelling a method](#cancelling-a-method)  
+    - [Testing environments](#testing-environments)  
+* [Topics](#topics)  
+    - [npe2](#npe2)  
+    - [Widget development](#widget-development)
+  
+## Questions
+
+### Testing plugins    
+Question: What are the best practices to test a plugin with multiple sequential steps?  
+e.g. Is it ok to rely on the "results" of a test to run the next test or should they all be fully independent?  
+
+Answer:  
+* Ideally, aim for unit testing.  
+* Tests should not pass/fail together.  
+* Use fixtures to provide a test with inputs, even if you have to make them up.  
+  https://docs.pytest.org/en/6.2.x/fixture.html 
+* Use mocks (mock-ups) to assert that specific calls are made, without necessarily caring about what happens after that call is made.  
+  https://docs.python.org/3/library/unittest.mock.html 
+
+*This is definitely an art form. It takes time. Be patient.*
+
+### Testing widgets
+Question: How do you test widgets, the napari viewer, and graphical user interfaces? How do you test Qt in general?
+
+Answer:
+* Try not to!
+* You should generally trust that a button click (for example) will call your callback and focus on testing that your callback does what it's supposed to do given that it gets called following some UI interaction.
+* However: If you have a scenario where you are actually creating a complicated widget directly in Qt, see `pytest-qt` for lots of tips, specifically `qtbot`.
+    - pytest-qt - [https://pytest-qt.readthedocs.io/en/latest/intro.html](https://pytest-qt.readthedocs.io/en/latest/intro.html)
+    - qtbot - [https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot](https://pytest-qt.readthedocs.io/en/latest/reference.html?highlight=qtbot#module-pytestqt.qtbot)
+* Oftentimes, this comes down to knowing and/or learning the Qt API really well.  
+
+### Better signal or slot documentation or functionality
+Question: How to create better signal/slot documentation/functionality as the user interacts with the core napari interface? For example, creating/editing/deleting a point or shape layer?
+
+Answer: 
+[https://napari.org/guides/stable/events_reference.html](https://napari.org/stable/guides/events_reference.html)  
+Granted, this is a work in progress. 
+
+### Creating or deleting a point or shape layer:  
+    Viewer.layers.events.inserted  
+    Viewer.layers.events.removed  
+    Viewer.layers.events.moved  
+    Viewer.layers.events.changed  
+    Viewer.layers.events.reordered  
+
+Editing a point or shape layer may be complicated, because it may be a pure NumPy object. 
```

Comment:
```suggestion
Getting an event when the user is editing the data inside a point or shape layer (outside of the GUI interface) is complicated, because the user will be directly editing the native array object.
```